iOS/Collection View Management /DiffableCollectionViewController.swift

```diff
@@ -0,0 +1,129 @@
+//
+
+import Foundation
+
+class newDataSource<SectionType: Hashable, ItemType: Hashable>: CollectionViewDataSource<SectionType, ItemType> {
+
+}
+
+class DiffableCollectionViewController<SectionType: Hashable, ItemType: Hashable, DataSource: newDataSource<SectionType, ItemType>>: ViewController, UICollectionViewDelegate {
+    
+    lazy var dataSource = DataSource.init(collectionView: self.collectionView)
+
+    @Published var selectedItems: [ItemType] = []
+
+    private var __selectedItems: [ItemType] {
+        return self.collectionView.indexPathsForSelectedItems?.compactMap({ ip in
+            return self.dataSource.itemIdentifier(for: ip)
+        }) ?? []
+    }
+
+    let collectionView: CollectionView
+
+    init(with collectionView: CollectionView) {
+        self.collectionView = collectionView
+        super.init()
+    }
+
+    required init?(coder aDecoder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func initializeViews() {
+        super.initializeViews()
+
+        self.collectionView.allowsMultipleSelection = true
+
+        self.view.addSubview(self.collectionView)
+        self.collectionView.delegate = self
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        Task {
+            await self.loadData()
+        }.add(to: self.taskPool)
+    }
+
+    override func viewDidLayoutSubviews() {
+        super.viewDidLayoutSubviews()
+
+        self.collectionView.expandToSuperviewSize()
+    }
+
+    @MainActor
+    private func loadData() async {
+
+        guard !Task.isCancelled else {
+            self.collectionView.animationView.stop()
+            return
+        }
+
+        let dataDictionary =  await self.retrieveDataForSnapshot()
+
+        let snapshot = self.getInitialSnapshot(with: dataDictionary)
+        await self.dataSource.apply(snapshot,
+                                    collectionView: self.collectionView,
+                                    animationCycle: self.getAnimationCycle())
+
+        self.collectionView.animationView.stop()
+    }
+
+    private func getInitialSnapshot(with dictionary: [SectionType: [ItemType]]) -> NSDiffableDataSourceSnapshot<SectionType, ItemType> {
+
+        var snapshot = self.dataSource.snapshot()
+        snapshot.deleteAllItems()
+
+        let allSections = Array(dictionary.keys)
+        snapshot.appendSections(allSections)
+
+        allSections.forEach { section in
+            if let items = dictionary[section] {
+                snapshot.appendItems(items, toSection: section)
+            }
+        }
+
+        return snapshot
+    }
+
+    // MARK: Overrides
+
+    // Used to capture and store any data needed for the snapshot
+    // Dictionary must include all SectionType's in order to be properly displayed
+    // Empty array may be returned for sections that dont have items.
+    func retrieveDataForSnapshot() async -> [SectionType: [ItemType]] {
```

Comment:
Notice here that we return a dictionary. No need to store an array of everything that we want to display. We can just use the datasource to access all our data. No need to keep another array up to date. I really like this approach.